File: src/main/java/com/model/annotation/XSecondsPastOrPresentImpl.java
package com.model.annotation;

import com.exception.OldTransactionException;

import javax.validation.ConstraintValidator;
import javax.validation.ConstraintValidatorContext;
import java.sql.Timestamp;
import java.util.Optional;
import java.util.concurrent.TimeUnit;

public class XSecondsPastOrPresentImpl  implements ConstraintValidator<XSecondsPastOrPresent, Timestamp> {

    private Integer seconds;

    @Override
    public void initialize(XSecondsPastOrPresent constraintAnnotation) {
        this.seconds = constraintAnnotation.seconds();
    }

    @Override
    public boolean isValid(Timestamp timestamp, ConstraintValidatorContext context) {
        Optional<Timestamp> optionalTimestamp = Optional.ofNullable(timestamp);

        long pastTime = TimeUnit.MILLISECONDS.toNanos(System.currentTimeMillis()) - TimeUnit.MILLISECONDS.toNanos(optionalTimestamp.orElse(new Timestamp(0)).getTime());
        long differenceInSeconds = TimeUnit.NANOSECONDS.toSeconds(pastTime);

        return (differenceInSeconds < seconds) ? true : false;
    }

}

File: src/main/kotlin/com/model/annotation/XSecondsPastOrPresentImpl.kt
package com.model.annotation

import mu.KotlinLogging
import javax.validation.ConstraintValidator
import java.sql.Timestamp
import javax.validation.ConstraintValidatorContext
import java.util.concurrent.TimeUnit

private val logger = KotlinLogging.logger {}

class XSecondsPastOrPresentImpl : ConstraintValidator<XSecondsPastOrPresent, Timestamp?> {
    private var seconds: Int = 60

    override fun initialize(constraintAnnotation: XSecondsPastOrPresent) {
        seconds = constraintAnnotation.seconds
    }

    override fun isValid(timestamp: Timestamp?, context: ConstraintValidatorContext): Boolean {
        val pastTime = TimeUnit.MILLISECONDS.toNanos(System.currentTimeMillis()) -
                TimeUnit.MILLISECONDS.toNanos((timestamp ?: Timestamp(0)).time)
        val differenceInSeconds = TimeUnit.NANOSECONDS.toSeconds(pastTime)

        logger.debug { "isValid return: differenceInSeconds: $differenceInSeconds seconds: $seconds" }
        return differenceInSeconds < seconds
    }
}